internal/pubsub/awspubsub/awspubsub.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2018 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package awspubsub provides an implementation of pubsub that uses AWS
+// SNS (Simple Notification Service) and SQS (Simple Queueing Service).
+//
+// It exposes the following types for As:
+// Topic: *sns.SNS
+// Subscription: *sqs.SQS
+package awspubsub
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sns"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"gocloud.dev/internal/pubsub"
+	"gocloud.dev/internal/pubsub/driver"
+)
+
+type topic struct {
+	mu     sync.Mutex
+	client *sns.SNS
+	arn    string
+}
+
+// TopicOptions will contain configuration for topics.
+type TopicOptions struct{}
+
+// OpenTopic opens the topic on AWS SNS for the given SNS client and topic ARN.
+func OpenTopic(ctx context.Context, client *sns.SNS, topicARN string, opts *TopicOptions) *pubsub.Topic {
+	dt := openTopic(ctx, client, topicARN)
+	return pubsub.NewTopic(dt)
+}
+
+// openTopic returns the driver for OpenTopic. This function exists so the test
+// harness can get the driver interface implementation if it needs to.
+func openTopic(ctx context.Context, client *sns.SNS, topicARN string) driver.Topic {
+	return &topic{client: client, arn: topicARN}
+}
+
+// SendBatch implements driver.Topic.SendBatch.
+func (t *topic) SendBatch(ctx context.Context, dms []*driver.Message) error {
+	for _, dm := range dms {
+		attrs := map[string]*sns.MessageAttributeValue{}
+		for k, v := range dm.Metadata {
+			attrs[k] = &sns.MessageAttributeValue{
+				DataType:    aws.String("String"),
+				StringValue: aws.String(v),
+			}
+		}
+		params := sns.PublishInput{
+			Message:           aws.String(string(dm.Body)),
+			MessageAttributes: attrs,
+			TopicArn:          &t.arn,
+		}
+		if err := t.send(&params); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *topic) send(params *sns.PublishInput) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	req, _ := t.client.PublishRequest(params)
```

Comment:
So AWS has no way to batch requests? Sad.

Reply:
I haven't found anything for batching sends, though it does appear to at least potentially receive in batches.